commercial/app/controllers/commercial/Multi.scala

```diff
@@ -21,94 +24,95 @@ class Multi(bestsellersAgent: BestsellersAgent,
   with implicits.Collections
   with implicits.Requests {
 
-  def renderMulti() = Action.async { implicit request =>
-    val requestedContent = request.getParameters("components")
+  private def multiSample(offerTypes: Seq[String], offerIds: Seq[Option[String]], segment: Segment): Future[Seq[Merchandise]] = {
+    val components = offerTypes zip offerIds
 
-    val slotIds = request.getParameters("slotIds")
+    val samples = components map {
+      case ("Book", Some(bookId))     => bestsellersAgent.getSpecificBooks(Seq(bookId))
+      case ("Book", None)             => Future.successful(bestsellersAgent.bestsellersTargetedAt(segment))
+      case ("Job", Some(jobId))       => Future.successful(jobsAgent.specificJobs(Seq(jobId)))
+      case ("Job", None)              => Future.successful(jobsAgent.jobsTargetedAt(segment))
+      case ("Masterclass", Some(masterclassId)) => Future.successful(masterclassAgent.specificMasterclasses(Seq(masterclassId)).filterNot(_.mainPicture.isEmpty))
+      case ("Masterclass", None)                => Future.successful(masterclassAgent.masterclassesTargetedAt(segment).filterNot(_.mainPicture.isEmpty))
+      case ("Soulmates", _)           => Future.successful {
+        (for {
+          woman <- SoulmatesAgent.womenAgent.sample().headOption
+          man <- SoulmatesAgent.menAgent.sample().headOption
+        } yield {
+          Seq(MemberPair(woman, man))
+        }).getOrElse {
+          Nil
+        }
+      }
+
+      case ("Travel", Some(travelId)) => Future.successful(travelOffersAgent.specificTravelOffers(Seq(travelId)))
+      case ("Travel", None)           => Future.successful(travelOffersAgent.offersTargetedAt(segment))
+      case _                          => Future.successful(Nil)
+    }
 
-    val componentsAndSpecificIds = requestedContent zip slotIds
+    Future.sequence(samples) map { realSamples =>
+        realSamples.map(_.headOption)
+    } map { realSamples =>
+        realSamples.flatten
+    }
+  }
+
+  def renderMulti() = Action.async { implicit request =>
+    val requestedContent = request.getParameters("components").map { component => component match {
+        case "books"  => "Book"
+        case "jobs"   => "Job"
+        case "travel" => "Travel"
+        case "masterclasses" => "Masterclass"
+        case "soulmates" => "Soulmates"
+        case _        => ""
+    }}
+
+    val slotIds = request.getParameters("slotIds").map { slotId => if (slotId.trim.isEmpty) None else Some(slotId) }
```

Comment:
That is also why I need to change to pattern matching in `multiSample` so that the `Some(String)` falls back to the `None` case (now I remember why I needed `getOrElse`).